torch/csrc/utils/tensor_new.cpp

```diff
@@ -100,6 +100,50 @@ static std::vector<int64_t> compute_sizes(PyObject* seq) {
   return sizes;
 }
 
+static ScalarType infer_scalar_type(PyObject *obj) {
+  if (PyFloat_Check(obj)) {
+    // this is always guaranteed to be a floating-point type, and makes it more
+    // convenient to write e.g. torch.tensor(0.) than torch.tensor(0., dtype=torch.Tensor.dtype).
+    return torch::tensor::get_default_tensor_type().scalarType();
+  }
+  if (THPUtils_checkLong(obj)) {
+    return ScalarType::Long;
+  }
+  if (PyBool_Check(obj)) {
+    // TODO: infer Bool when we have Bool ScalarType
+    return ScalarType::Byte;
+  }
+  if (THPVariable_Check(obj)) {
+    auto var = reinterpret_cast<THPVariable*>(obj)->cdata;
+    return var.type().scalarType();
+  }
+#ifdef WITH_NUMPY
+  if (PyArray_Check(obj)) {
+    auto array = (PyArrayObject*)obj;
+    return numpy_dtype_to_aten(PyArray_TYPE(array));
+  }
+#endif
+  if (PySequence_Check(obj)) {
+    ScalarType scalarType = ScalarType::NumOptions;
```

Comment:
This is fine, but you can also use:

```
at::optional<ScalarType> scalarType;
```

i.e. use nullopt as the default value instead of NumOptions.